qiskit/pulse/schedule.py

```diff
@@ -322,13 +328,46 @@ def draw(self, dt: float = 1, style: Optional['SchedStyle'] = None,
                                           interactive=interactive, table=table,
                                           label=label, framechange=framechange)
 
-    def __add__(self, schedule: ScheduleComponent) -> 'Schedule':
-        """Return a new schedule with `schedule` inserted within `self` at `start_time`."""
-        return self.append(schedule)
+    def __eq__(self, other: ScheduleComponent) -> bool:
+        """Test if two ScheduleComponents are equal.
+
+        Equality is checked by verifying there is an equal instruction at every time
+        in `other` for every instruction in this Schedule.
+
+        Warning: This does not check for logical equivalencly. Ie.,
+            ```python
+            >>> (Delay(10)(DriveChannel(0)) + Delay(10)(DriveChannel(0)) ==
+                 Delay(20)(DriveChannel(0)))
+            False
+            ```
+        """
+        channels = set(self.channels)
+        other_channels = set(other.channels)
+
+        # first check channels are the same
+        if channels != other_channels:
+            return False
+
+        # then verify same number of instructions in each
+        instructions = self.instructions
+        other_instructions = other.instructions
+        if len(instructions) != len(other_instructions):
+            return False
+
+        # finally check each instruction in `other` is in this schedule
+        for inst in other_instructions:
+            if inst not in instructions:
```

Comment:
`in` for a tuple will generally be slow (O(n)), I believe. I think that we should be able to improve this and the above issue at the same time by comparing each inst:
```
for idx, inst in enumerate(other_instructions):
    if instructions[i] != inst or instructions[i].time != inst.time:  # `.time` is not a thing, but thats the idea...
        return False
```

Reply:
This relies on the order of `instructions` being consistent between schedules, which could be a problem since the semantic meaning of the pulse is encoded in its time and Command. This should be fine now as `instructions` is sorted. In the future, it shouldn't even be a question as instructions will be relatively ordered.